pandora-server-directory/src/database/mongoDb.ts

```diff
@@ -518,7 +520,137 @@ export default class MongoDatabase implements PandoraDatabase {
+
+		// Some stats to collect throughout migration
+		const startTime = Date.now();
+		let success = true;
+		let totalCount = 0;
+		let changeCount = 0;
+
+		// A generic function that processes a single collection, applying a schema to each document in the collection
+		const performMigration = async <T extends object>(name: string, collection: Collection<T>, schema: ZodType<T, ZodTypeDef, unknown>): Promise<void> => {
+			migrationLog.info(`Processing ${name}...`);
+
+			for await (const originalData of collection.find().stream()) {
+				totalCount++;
+				const documentId: ObjectId = originalData._id;
+				Assert(documentId instanceof ObjectId);
+
+				const originalDataWithoutType: Record<string, unknown> = originalData;
+				// Parse the data using schema (this is the main bit of migration)
+				const parsedData = schema.safeParse(originalData);
+				if (!parsedData.success) {
+					success = false;
+					migrationLog.error(`Failed to migrate ${name} document ${documentId.toHexString()}:\n`, parsedData.error);
+					continue;
+				}
+
+				if (isEqual(omit(originalData, '_id'), omit(parsedData.data, '_id')))
+					continue;
+
+				// If the parse result is not equal, the data needs migration
+				changeCount++;
+
+				// Collect updated properties and keys to delete
+				const update: Partial<T> = {};
+				const keysToDelete = new Set<string>();
+				for (const [key, value] of KnownObject.entries(parsedData.data)) {
+					if (key === '_id')
+						continue;
+
+					if (value !== undefined && !isEqual(originalDataWithoutType[key as string], value)) {
+						update[key] = value;
+					}
+				}
+				for (const key of Object.keys(originalData)) {
+					if (key === '_id')
+						continue;
+
+					if (parsedData.data[key as keyof T] === undefined) {
+						keysToDelete.add(key);
+					}
+				}
+
+				// Generate a diff for manual review
+				const diff = diffString(omit(originalData, '_id'), omit(parsedData.data, '_id'), { color: false });
+				migrationLog.verbose(
+					`Migrating ${name} document ${documentId.toHexString()}...\n`,
+					`Updates keys: ${Object.keys(update).join(', ')}\n`,
+					`Removed keys: ${Array.from(keysToDelete).join(', ')}\n`,
+					`Diff:\n`,
+					diff,
+				);
+
+				if (dryRun)
+					continue;
+
+				// Actually perform the update
+				const { matchedCount } = await collection.updateOne(
+					// @ts-expect-error: This works; typechecking is just broken for some reason
+					{ _id: documentId },
+					{
+						$set: update,
+						$unset: ArrayToRecordKeys(Array.from(keysToDelete), true),
+					},
+				);
+				if (matchedCount !== 1) {
+					success = false;
+					migrationLog.error(`Failed to migrate ${name} document ${documentId.toHexString()}: Update matched count is ${matchedCount}`);
+				}
+			}
+		};
+
+		// Perform migration on all collections
+		await performMigration<DatabaseAccountWithSecure>(
+			'accounts',
+			this._accounts,
+			DatabaseAccountWithSecureSchema,
```

Comment:
my issue with this approach is that it can only do what the load time check already does in some of the cases

what this lacks is true migration when we need to transform the old structure to a new one, like if we want to rename `chatroom` to `space`, or any other more complex type transformation

a more versetile aproach could be to create a collection class, that can do the loading migration all the time, so we do not have to sprinkle it everywhere, and accept 2 optional parameter, old schema, and transformation function

```ts
class ValidatedCollection<TSchema extends ZodTypeAny, TOldSchema extends ZodTypeAny = TSchema> {
	constructor(
		key: string,
		indexes: (IndexDescription & { name: string; })[],
		schema: TSchema,
		oldSchema?: TOldSchema,
		migration?: (old: z.infer<TOldSchema>) => z.infer<TSchema>,
	) {
	}
}
```

Reply:
Of course there are many possibilities of approaching this, many more advanced than this.
I maybe wasn't fully clear on the goals of _this_ implementation, those are two:
- Ability to do a "dry run" on the whole database to see what the schema changes actually do on real data - to see that they are not invalidating something that should be valid.
- Ability to sweep through the whole database AoT style instead of only when specific value is loaded to cleanup old, potentially inactive entries.
Yes, we can easily add a pre-/post-parse hooks for manipulating either old data before schema check or data after schema check if the needed migration is not only about schemas.